### PR TITLE
Upgrade to ZIO 2.0

### PR DESCRIPTION
**Pull request by justcoon**
Upgrade to ZIO 2.0  - #145

### CODE REVIEW

File: src/main/scala/zio/akka/cluster/Cluster.scala [hunk 1]
```diff
@@ -3,20 +3,20 @@ package zio.akka.cluster
...
-import zio.{ Has, Queue, Runtime, Task, ZIO }
+import zio.{ Queue, Runtime, Task, ZIO }
 
 object Cluster {
 
-  private val cluster: ZIO[Has[ActorSystem], Throwable, akka.cluster.Cluster] =
+  private val cluster: ZIO[ActorSystem, Throwable, akka.cluster.Cluster] =
     for {
-      actorSystem <- ZIO.access[Has[ActorSystem]](_.get)
+      actorSystem <- ZIO.environmentWith[ActorSystem](_.get)
```
Comment (ghostdogpr) [inline src/main/scala/zio/akka/cluster/Cluster.scala]: Can we use `ZIO.service` to simplify this? (here and elsewhere)

File: src/main/scala/zio/akka/cluster/sharding/Sharding.scala [hunk 2]
```diff
@@ -37,14 +37,14 @@ object Sharding {
    * @param askTimeout     a finite duration specifying how long an ask is allowed to wait for an entity to respond
    * @return a [[Sharding]] object that can be used to send messages to sharded entities
    */
-  def start[R <: Has[_], Msg, State: Tag](
+  def start[R <: Any, Msg, State: Tag](
```
Comment (ghostdogpr) [inline src/main/scala/zio/akka/cluster/sharding/Sharding.scala]: I think we can remove this constraint now

File: src/main/scala/zio/akka/cluster/sharding/Sharding.scala [hunk 3]
```diff
@@ -129,7 +129,7 @@ object Sharding {
       )
   }
 
-  private[sharding] class ShardEntity[R <: Has[_], Msg, State: Tag](rts: Runtime[R])(
+  private[sharding] class ShardEntity[R <: Any, Msg, State: Tag](rts: Runtime[R])(
```
Comment (ghostdogpr) [inline src/main/scala/zio/akka/cluster/sharding/Sharding.scala]: Here too

File: src/test/scala/zio/akka/cluster/sharding/ShardingSpec.scala [hunk 4]
```diff
@@ -183,8 +191,8 @@ object ShardingSpec extends DefaultRunnableSpec {
                 p2        <- Promise.make[Nothing, Unit]
                 onMessage1 = (_: String) => p1.succeed(()).unit
                 onMessage2 = (_: String) => p2.succeed(()).unit
-                sharding1 <- Sharding.start(shardName, onMessage1).provideLayer(ZLayer.succeedMany(a1))
-                _         <- Sharding.start(shardName, onMessage2).provideLayer(ZLayer.succeedMany(a2))
+                sharding1 <- Sharding.start(shardName, onMessage1).provideLayer(ZLayer.succeedEnvironment(a1))
```
Comment (ghostdogpr) [inline src/test/scala/zio/akka/cluster/sharding/ShardingSpec.scala]: Can we simply use `succeed`?

**Review by ghostdogpr — COMMENTED** [review]
Looks great! My main suggestion is to simplify a bunch of `environment` with `service`, then it should be good to go.

I created a `series/2.x` branch to merge this into. Also, can you solve the conflict with that branch?

File: src/test/scala/zio/akka/cluster/sharding/ShardingSpec.scala (hunk 4 above)
Comment (justcoon) [reply src/test/scala/zio/akka/cluster/sharding/ShardingSpec.scala]: not sure what you mean by succeed,
but I changed it to
```
Sharding.start(shardName, onMessage1).provideEnvironment(a1)
```

File: README.md [hunk 5]
```diff
@@ -176,14 +176,14 @@ This library wraps messages inside of a `zio.akka.cluster.sharding.MessageEnvelo
...
 
-val actorSystem: ZLayer[Any, Throwable, Has[ActorSystem]] =
-  ZLayer.fromManaged(Managed.make(Task(ActorSystem("Test")))(sys => Task.fromFuture(_ => sys.terminate()).either))
+val actorSystem: ZLayer[Any, Throwable, ActorSystem] =
+  ZLayer.fromManaged(Managed.acquireReleaseWith(Task(ActorSystem("Test")))(sys => Task.fromFuture(_ => sys.terminate()).either))
 
 val behavior: String => ZIO[Entity[Int], Nothing, Unit] = {
-  case "+" => ZIO.accessM[Entity[Int]](_.get.state.update(x => Some(x.getOrElse(0) + 1)))
-  case "-" => ZIO.accessM[Entity[Int]](_.get.state.update(x => Some(x.getOrElse(0) - 1)))
+  case "+" => ZIO.service[Entity[Int]].flatMap(_.state.update(x => Some(x.getOrElse(0) + 1)))
```
Comment (ghostdogpr) [inline README.md]: You can do `serviceWithZIO` to simplify this slightly

File: src/main/scala/zio/akka/cluster/sharding/package.scala [hunk 6]
```diff
@@ -21,19 +21,19 @@ package object sharding {
     }
 
     def replyToSender[State: Tag, R](msg: R): ZIO[Entity[State], Throwable, Unit]         =
-      ZIO.accessM[Entity[State]](_.get.replyToSender(msg))
+      ZIO.service[Entity[State]].flatMap(_.replyToSender(msg))
```
Comment (ghostdogpr) [inline src/main/scala/zio/akka/cluster/sharding/package.scala]: Here as well (and some more below)

File: src/test/scala/zio/akka/cluster/sharding/ShardingSpec.scala [hunk 7]
```diff
@@ -3,14 +3,16 @@ package zio.akka.cluster.sharding
...
+import zio.Clock
+
 import zio.test.Assertion._
 import zio.test._
-import zio.test.environment.TestEnvironment
-import zio.{ ExecutionStrategy, Has, Managed, Promise, Task, UIO, ZIO, ZLayer }
+import zio.test.TestEnvironment
+import zio.{ ExecutionStrategy, Managed, Promise, Task, UIO, ZIO, ZLayer }
+import zio._
+import zio.test.ZIOSpecDefault
```
Comment (ghostdogpr) [inline src/test/scala/zio/akka/cluster/sharding/ShardingSpec.scala]: minor: can you sort imports?

File: src/test/scala/zio/akka/cluster/sharding/ShardingSpec.scala [hunk 8]
```diff
@@ -75,23 +77,23 @@ object ShardingSpec extends DefaultRunnableSpec {
...
       },
-      testM("gather state") {
+      test("gather state") {
         assertM(
           for {
             p         <- Promise.make[Nothing, Boolean]
             onMessage  = (_: String) =>
                            for {
-                             state    <- ZIO.access[Entity[Int]](_.get.state)
+                             state    <- ZIO.service[Entity[Int]].map(_.state)
```
Comment (ghostdogpr) [inline src/test/scala/zio/akka/cluster/sharding/ShardingSpec.scala]:
```suggestion
                             state    <- ZIO.serviceWith[Entity[Int]](_.state)
```

**Review by ghostdogpr — COMMENTED** [review]
LGTM! Just a tiny remaining one.

**Pull request merged**